### PR TITLE
Visualization Keys and Labels: Initial commit in class

### PR DESCRIPTION
**Pull request by NikitaThomas**
Implementing Issue #9 

- This is my initial commit during class. File is a work in progress. Included are only the raw_input prompts.

### CODE REVIEW

File: hs628_tools/Visualization.py [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+        Q2= input("What do you want to title the graph?"+"\n")
+        Q3 = input("What is the predictor(dependent) variable name?:"+"\n")
+        Q4 = input("Name of x axis:")
+        Q5 = input("Name of y axis:") 
+        Q6 = input("Choose color for actual y points:"+"\n")
+        Q7 = input("Choose color for line:"+"\n")
+        model_metrics(X_train, X_test, y_train, y_test)
+        return plot_model(X_train, X_test, y_train, y_test,Q2,Q3,Q4,Q5,Q6,Q7)
+
+##### MAIN #####
```
Comment (omidkj) [inline hs628_tools/Visualization.py]: Your code is really great, with attention to details, and well structured. 
As we discussed in the class, you can define a method (function) for Q1 and take it out of your main program. I'd also use the test part in __main__:

```python
def type_of_plot_question():
    result = input("What type of plot output? (Enter #)"+"\n"+"\n"+"1.Countplot"+"\n"+"2.Scatterplot"+"\n"+"3.Boxplot"+"\n"+"4.Simple Linear Regression Model"+"\n"+"5.Multiple Linear Regression Model"+"\n")
   return result

if __name__ == "__main__":
    Xtrain, Xtest, ytrain, ytest = train_test_split(X,Y, test_size = .5) 
    #This assumes that Issue #11 Test.py module works and outputs the test ratio. 
    Xtrain = np.array(Xtrain)
    ytrain = np.array(ytrain)
    Xtest = np.array(Xtest)
    ytest = np.array(ytest)
    type_of_plot = type_of_plot_question()
    question(data,Xtrain,Xtest,ytrain,ytest, type_of_plot )
```

I think you could also create a class for your solution, so your code could be used in other people's tools.

**Comment by rohitchadaram** [comment]
@NikitaThomas I have added some of comments on @haleyhowe  pull request work please feel free to address them, as I think you and Haley worked together on this. Overall I think this is a great idea, well executed to avoid any hard-coding in the plots. There are small nuances like using elif when you have multiple condition checks and also the unnecessary use of pass when you can tweak your if condition to negate the condition checking. Apart from those great idea and great job guys.

File: hs628_tools/Visualization.py [hunk 2]
```diff
@@ -0,0 +1,294 @@
+import matplotlib as mpl
+import matplotlib.pyplot as plt
+import numpy as np
+import seaborn as sns
+import pandas as pd
```
Comment (EHWUSF) [inline hs628_tools/Visualization.py]: Temporarily using pandas to simpliy reading in data for testing could be ok, but as discussed in the written assignment and in class, the final module was not supposed to rely on pandas.

File: hs628_tools/Visualization.py [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+import sys 
+import plotly.plotly as py
```
Comment (EHWUSF) [inline hs628_tools/Visualization.py]: This script does not currently run for me, so I can't fully test this, but what I've read of plotly now seems to indicate that this code would only run within a live ipython notebook.  (There are ways to use plotly that are at least compatible with offline notebooks, but this module does not appear to use that structure.)  Have you guys tried running this from within a regular python script or checking it in an offline notebook?   A module that can only be used within an ipython notebook still has value, and wasn't prohibited by the assignment, but such constraints, and proper manner of use, need to be very clearly documented.

File: hs628_tools/Visualization.py [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+import plotly.graph_objs as go
+#plotly.tools.set_credentials_file(username='hhowell', api_key='BRP1aHhfTHwkqnvalVQL')
```
Comment (EHWUSF) [inline hs628_tools/Visualization.py]: The substance of this commit, and the commit message immediately preceding this commit, suggest that this commit represents adding Haley's code; if that is the case then the commit message here should have made it clear.  Even better would be if it were a merge commit, such as would result if one of you had submitted a pull request to the other one's feature branch on their fork, and that PR was then merged into that feature branch (which is what I thought we had talked about), and then that feature branch was the basis for a PR into the main repository.  That way, that feature branch would carry all of the relevant git history for both authors along with it, instead of a bunch of it being collapsed into a single commit by just the one author.  

Since I'm already a lot of the way through commenting this amalgamated file, for the remaining cleanup work, which you'll both want to do respective parts of, it would probably be best for @haleyhowe  to branch off of @NikitaThomas feature branch that is the basis for this PR, then for each of you to make your respective changes, then for Haley to submit a PR to Nakita's feature branch, which Nakita would then merge into her branch to have those changes become part of this PR.  It would probably be worthwhile to step through that in class, but if we don't have time in class then we should do it after or schedule time over zoom.

File: hs628_tools/Visualization.py [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+from sklearn import datasets, linear_model
+from sklearn.model_selection import train_test_split
+from sklearn.metrics import mean_squared_error,mean_absolute_error, r2_score
+###This is importing a class from Issue #11 Test module. We used the ratio function to determine the test size ratio for the train_test_split function. 
+#from Test import data_split as ds
+
+
+###Bar Plot-Count
+#This function counts all instances in a variable and creates a barplot to display the results 
+#@param: The variable to be counted, the x_axis label, title of the graph
```
Comment (EHWUSF) [inline hs628_tools/Visualization.py]: Although I've seen this kind of @param documentation syntax used with other languages, I'm not familiar with it for python, and PyCharm does not seem to recognize it either.  Best to stick with a syntax that is standard for the language, there is a link or two in one of the class announcements.  And that preceding comment should instead be in a docstring.

File: hs628_tools/Visualization.py [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+#@return: A count bar plot of the variable inputted 
+def count_and_print(variable, x_label, title):
```
Comment (EHWUSF) [inline hs628_tools/Visualization.py]: ```variable``` is a rather nondescript name for a variable.  And although we could reasonably infer that ```x_label``` and ```title``` should be strings, it should be clearly identified (by the name or in the docstring) what type this other parameter should be.

File: hs628_tools/Visualization.py [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+    def count(variable):
```
Comment (EHWUSF) [inline hs628_tools/Visualization.py]: It can be fine to keep the outer namespace cleaner/leaner by doing imports and defining helper functions within another function that that's the only place they are used, but then you should still follow standard rules of well organized python code-- imports first, then class & function definitions, then other code.

File: hs628_tools/Visualization.py [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+        new_dic = {}
+        c = (len(variable)-1)
+        for index, value in enumerate(variable):
+            count = 0
+            if((value in new_dic) == True):
```
Comment (EHWUSF) [inline hs628_tools/Visualization.py]: I think someone else already noted the ```pass``` here was slightly awkward and could be avoided by inverting the logic of your test or changing the block structure.  There is also no need for the ```== True``` 

In a loop like this, with no code following the if block, it can often be slightly cleaner code to replace 
```python
    if value not in new_dic:
        for index2, value2 in enumerate(variable):
        ...
```
with 
```python
    if value in new_dic:
        continue
    for index2, value2 in enumerate(variable):
```

File: hs628_tools/Visualization.py [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+                pass
+            else:
+                for index2, value2 in enumerate(variable):
+                    if value == value2:
+                        count = count + 1
+                    if index2 == (len(variable)-1):
```
Comment (EHWUSF) [inline hs628_tools/Visualization.py]: This is looking rather strange to me, the operation of this function should be clarified with comments.

File: hs628_tools/Visualization.py [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+
+    variable = variable
+
+    count_dictionary = count(variable)
+    from collections import OrderedDict
+    from operator import itemgetter
+
+    count_ordered = (OrderedDict(sorted(count_dictionary.items(), key = itemgetter(0), reverse = False)))
+
+    def key_list(list):
```
Comment (EHWUSF) [inline hs628_tools/Visualization.py]: you shouldn't shadow (overwrite) a keyword or standard type like list, choose a different variable name.

File: hs628_tools/Visualization.py [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+        keys_list= []
+        for key in list.keys():
```
Comment (EHWUSF) [inline hs628_tools/Visualization.py]: python lists don't have a ```keys()``` method, so ```list``` is a doubly poor variable name here.

File: hs628_tools/Visualization.py [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
+            keys_list.append(key)
+        return keys_list
+
+    def value_list(list):
```
Comment (EHWUSF) [inline hs628_tools/Visualization.py]: similar comments as for ```key_list```

File: hs628_tools/Visualization.py [hunk 13]
```diff
@@ -0,0 +1,294 @@
...
+        values_list= []
+        for value in list.values():
+            values_list.append(value)
+        return values_list
+    keys = key_list(count_ordered)
```
Comment (EHWUSF) [inline hs628_tools/Visualization.py]: I'm not familiar with plotly, but if for some reason you can't just use ```x=count_ordered.keys()``` and similar for values(), down below, then at least you should be able to simply do ```keys = list(count_ordered.keys())``` and similar for values(), without needing to make your own special functions.

File: hs628_tools/Visualization.py [hunk 14]
```diff
@@ -0,0 +1,294 @@
...
+    values = value_list(count_ordered)
+#Plot the variable
+    f1 = (go.Bar(
```
Comment (EHWUSF) [inline hs628_tools/Visualization.py]: This ```f1 =``` line looks like it should all fit on a single line just fine.

File: hs628_tools/Visualization.py [hunk 15]
```diff
@@ -0,0 +1,294 @@
...
+                x=keys,
+                y=values
+        )
+           )
+    layout = go.Layout(
```
Comment (EHWUSF) [inline hs628_tools/Visualization.py]: this ```layout =``` line may be too long for a single line, but the current formatting seems too disperse.

File: hs628_tools/Visualization.py [hunk 16]
```diff
@@ -0,0 +1,294 @@
...
+            )
+    )
+
+    
+    data = [f1]
+    fig = go.Figure(data=data,layout=layout)
+    return py.iplot(fig, filename='basic-bar')
+
+
+###Scatter Plot
```
Comment (EHWUSF) [inline hs628_tools/Visualization.py]: Better to have a docstring than this comment.

File: hs628_tools/Visualization.py [hunk 17]
```diff
@@ -0,0 +1,294 @@
...
+                color='#7f7f7f'
+                )
+            )
+    )
+    data = [trace1]
+    fig = go.Figure(data=data,layout=layout)
+    return py.iplot(fig, filename='scatter-plot-with-colorscale');
+
+###BoxPlot
+#This function creates a box plot of the two variables the user decides to examine
```
Comment (EHWUSF) [inline hs628_tools/Visualization.py]: Same docstring comments as above.

File: hs628_tools/Visualization.py [hunk 18]
```diff
@@ -0,0 +1,294 @@
...
+#X = sub_data.iloc[:,0:3]
+#Y = sub_data['readmission_30d']
+
+Xtrain, Xtest, ytrain, ytest = train_test_split(X,Y, test_size = .5) #This assumes that Issue #11 Test.py module works and outputs the test ratio. 
+Xtrain = np.array(Xtrain)
+ytrain = np.array(ytrain)
+Xtest = np.array(Xtest)
+ytest = np.array(ytest)
+
+Q1= input("What type of plot output? (Enter #)"+"\n"+"\n"+"1.Countplot"+"\n"+"2.Scatterplot"+"\n"+"3.Boxplot"+"\n"+"4.Simple Linear Regression Model"+"\n"+"5.Multiple Linear Regression Model"+"\n")
```
Comment (EHWUSF) [inline hs628_tools/Visualization.py]: This line in particular makes this all more like a script than a module.  The mere import of a module generally should not trigger a requirement for user interaction, which this line would do.  You should instead have such user interaction get triggered by calling a function from your module.

**Review by EHWUSF — COMMENTED** [review]
Please add some instructions to the documentation about how the final version is to be used, and if there are any special restrictions on the environment in which it can be used.

File: hs628_tools/Visualization.py (hunk 4 above)
Comment (EHWUSF) [reply hs628_tools/Visualization.py]: Hmm, now I see, though, that Haley's final version looks like it has a couple more commits cleaning up a couple of the issues with this final version.  But it looks like her final version also has some other stray code in it that's been cleaned up here, so it's not immediatlely clear which one would be the better basis for final cleanup, you guys choose.

File: hs628_tools/Visualization.py [hunk 19]
```diff
@@ -0,0 +1,294 @@
...
+
+##### MAIN #####
+
+##Note these four lines of code are meant to test this file 
+#data = pd.read_csv('data.csv')
+#sub_data = data.iloc[1:10,1:10]
+#X = sub_data.iloc[:,0:3]
+#Y = sub_data['readmission_30d']
+
+Xtrain, Xtest, ytrain, ytest = train_test_split(X,Y, test_size = .5) #This assumes that Issue #11 Test.py module works and outputs the test ratio.
```
[the remark this answers is not in the document]
Comment (NikitaThomas) [reply hs628_tools/Visualization.py:283]: Hi Nirvesh, I just chose .5 as an arbitrary number to test out my code but ideally the user will be using your module to find the right ratio.
